lib/Conversion/GPUXToLLVM/GPUXToLLVMPass.cpp

```diff
@@ -0,0 +1,721 @@
+
+      auto typeSize = param.first;
+
+      mlir::Value range =
+          rewriter.create<mlir::LLVM::UndefOp>(loc, llvmRangeType);
+      range = rewriter.create<mlir::LLVM::InsertValueOp>(loc, range, ptr, 0);
+      range =
+          rewriter.create<mlir::LLVM::InsertValueOp>(loc, range, typeSize, 1);
+
+      paramsArray = rewriter.create<mlir::LLVM::InsertValueOp>(loc, paramsArray,
+                                                               range, i);
+    }
+
+    auto nullPtr = rewriter.create<mlir::LLVM::NullOp>(loc, llvmPointerType);
+    auto nullRange = [&]() {
+      auto zero = rewriter.create<mlir::LLVM::ConstantOp>(
+          loc, llvmIndexType, rewriter.getIntegerAttr(llvmIndexType, 0));
+      mlir::Value range =
+          rewriter.create<mlir::LLVM::UndefOp>(loc, llvmRangeType);
+      range =
+          rewriter.create<mlir::LLVM::InsertValueOp>(loc, range, nullPtr, 0);
+      range = rewriter.create<mlir::LLVM::InsertValueOp>(loc, range, zero, 1);
+      return range;
+    }();
+    paramsArray = rewriter.create<mlir::LLVM::InsertValueOp>(
+        loc, paramsArray, nullRange, paramsCount);
+    rewriter.create<mlir::LLVM::StoreOp>(loc, paramsArray, paramsArrayPtr);
+
+    auto paramsArrayVoidPtr = rewriter.create<mlir::LLVM::BitcastOp>(
+        loc, llvmRangePointerType, paramsArrayPtr);
+    auto nullpointer =
+        rewriter.create<mlir::LLVM::NullOp>(loc, llvmPointerPointerType);
+    launchKernelCallBuilder.create(
+        loc, rewriter,
+        {stream, function->getResult(0), adaptor.getGridSizeX(),
+         adaptor.getGridSizeY(), adaptor.getGridSizeZ(),
+         adaptor.getBlockSizeX(), adaptor.getBlockSizeY(),
+         adaptor.getBlockSizeZ(), adaptor.getDynamicSharedMemorySize(),
+         paramsArrayVoidPtr,
+         /*extra=*/nullpointer});
+
+    if (launchOp.getAsyncToken()) {
+      // Async launch: make dependent ops use the same stream.
+      rewriter.replaceOp(launchOp, {stream});
+    } else {
+      // Synchronize with host and destroy stream. This must be the stream
+      // created above (with no other uses) because we check that the
+      // synchronous version does not have any async dependencies.
+      waitCallBuilder.create(loc, rewriter, stream);
+      rewriter.eraseOp(launchOp);
+    }
+    moduleUnloadCallBuilder.create(loc, rewriter, module->getResult(0));
+
+    return mlir::success();
+  }
+}; // namespace
+
+/// A rewrite pattern to convert gpu.memcpy operations into a GPU runtime
+/// call. Currently it supports CUDA and ROCm (HIP).
+class ConvertMemcpyOpToGpuRuntimeCallPattern
+    : public ConvertOpToGpuRuntimeCallPattern<gpux::MemcpyOp> {
+public:
+  ConvertMemcpyOpToGpuRuntimeCallPattern(mlir::LLVMTypeConverter &typeConverter)
+      : ConvertOpToGpuRuntimeCallPattern<gpux::MemcpyOp>(typeConverter) {}
+
+private:
+  mlir::LogicalResult
+  matchAndRewrite(gpux::MemcpyOp memcpyOp, OpAdaptor adaptor,
```

Comment:
Interesting. Looks like this does not cause compile error. Missing implementation but not called from anywhere in the code.